### ISSUE TITLE
find_package catkin should not allow finding other packages as components

### ISSUE THREAD

**Issue by tkruse** [ros/catkin#255]
currently find_package(catkin components roscpp)
seems to be the equivalents of
find_package(catkin)
find_package(roscpp)
with added functionality of collecting the variable values into catkin_ prefixed variables.

However that is in violation of cmake's find_package documentation, where in this case roscpp should be a component specific to catkin, which it isn't. 

Similarly, later using
target_link_libraries(foo ${catkin_LIBRARIES})
For a package that does not use any cakin libraries (if there were any) is confusing. Also it is confusing that the variable ${catkin_LIBRARIES} contains other things than those catkin itself depends on.

Instead a macro could make all this explicit:

find_package_set(target1_dependencies roscpp)
target_link_libraries(target1  ${target1_dependencies_LIBRARIES})
find_package_set(target2_dependencies roscpp std_msgs)
target_link_libraries(target2  ${target2_dependencies_LIBRARIES})

With the current approach, lazy developers will use the same dependency set for both target1 and target2 in the example above, because the current approach does not allow to have both the convenience and separation of variables.

**Comment by tfoote** [comment]
From meeting discussion: the catkin COMPONENTS keeps sorting.  It's worth keeping the shorter user markup and automatically sorting.  We can expose the sorting method to the power user.  It's a little different but powerful.  For the 2nd use case you can use the component LIBRARIES directly but need to sort.  

To resolve this ticket we should expose the sort functionality.

**Comment by jbohren** [comment]
If we don't want to abuse `find_package(catkin)` then why not do something like:

```
find_package(catkin)
catkin_find_packages(pkg1 pkg2 pkg3)
```

And only after the second call will `catkin_LIBRARIES`, `catkin_INCLUDE_DIRS`, etc be defined. 

Then maybe we could even add something like `catkin_find_packages_from_xml()` which will automatically parse the `package.xml` file and make it so you don't have to declare the deps in both the `package.xml` and `CMakeLists.txt` files.

**Comment by dirk-thomas** [comment]
This topic has been discussed a lot on the SIG buildsystem mailing list . There are several very valid reasons why the dependencies are listed in both the package.xml and CMakeLists.txt files separately. As long as these are not disproved the current implementation will not get changed.

**Comment by tfoote** [comment]
We talked through this a little bit and there's some significant reasons for using the component based approach.  The current usage provides a significant feature which is that it properly sorts the libs and include dirs to enable our overlay system.  In our discussion we did believe that should be exposed separately ticketed in #259 . 

With respect to being a "standard" find_package usage.  It follows the standard find_package expectation that PACKAGENAME_LIBRARIES PACKAGENAME_\* are set for the package name.  And it's common practice to use the components, such as boost.  It's a little different as the components are from a federated system and not an integrated release, but as they need catkin to be able to chain in this way keeping them as catkin "components" seems reasonable.
